blog/2021-12-31-kubevela订购aws-s3中间过程演示.md

```diff
@@ -0,0 +1,467 @@
+***kubevela订购aws s3中间过程演示***
+
+------
+
+> 本文介绍如何通过Kubevela订购aws s3 bucket，演示kubevela在订购aws s3过程中，涉及到的kubevela application示例，以及kubevela在订购s3时，terraform-controller处理configuration的中间状态。对于在阅读kubevela terraform-controller代码时，本文可以作为参考，加深理解源码。
+
+# 创建aws-s3
+
+## 示例介绍
+
+使用官方aws-s3 comp，订购aws s3 bucket。通过观察中间过程，深入理解vela实现订购云资源的过程。可以对照这更好的read terraform controller代码。
+
+![architecture.jpg](https://github.com/oam-dev/terraform-controller/blob/v0.2.13/docs/resources/architecture.jpg?raw=true)
+
+## 演示步骤
+
+
+
+### 环境配置
+
+本示例演示环境是minikube+kubevela+terraform-controller，其版本信息如下：
+
+| 组件                 | 版本                                                         |
+| -------------------- | ------------------------------------------------------------ |
+| minikube             | minikube version: v1.23.2<br/>commit: 0a0ad764652082477c00d51d2475284b5d39ceed |
+| kubevela             | Version: v1.2.0-beta.2<br/>GitRevision: git-ad4b446<br/>GolangVersion: go1.16.10 |
+| terraform-controller | oamdev/terraform-controller:0.2.14                           |
+
+> 安装minikube、kubevela、terraform controller以及terraform-aws addon过程省略。
+
+### 定义配置provider
+
+***provider模板如下***：
+
+```
+apiVersion: terraform.core.oam.dev/v1beta1
+kind: Provider
+metadata:
+  name: aws
+  namespace: default
+spec:
+  credentials:
+    secretRef:
+      key: credentials
+      name: aws-account-creds
+      namespace: vela-system
+    source: Secret
+  provider: aws
+  region: us-east-2
+```
+
+provider使用secret aws-account-creds，其存储了terraform访问aws api所需的ak/sk信息：
+
+```
+apiVersion: v1
+data:
+  credentials: xxx
+kind: Secret
+metadata:
+  name: aws-account-creds
+  namespace: vela-system
+
+>>>base64 decode:
+awsAccessKeyID: xxx
+awsSecretAccessKey: xxx
+```
+
+> 不同云提供商认证方式不同，aws采用ak/sk方式，其他云请参考官方文档
+
+截止到kubevela 最新的1.2版本为止，provider支持如下，***且都是hardcode编码的***：
+
+```
+const (
+	alibaba CloudProvider = "alibaba"
+	aws     CloudProvider = "aws"
+	gcp     CloudProvider = "gcp"
+	azure   CloudProvider = "azure"
+	vsphere CloudProvider = "vsphere"
+	ec      CloudProvider = "ec"
+	ucloud  CloudProvider = "ucloud"
+)
+```
+
+### 定义application
+
+```
+apiVersion: core.oam.dev/v1beta1
+kind: Application
+metadata:
+  name: sml-aws-s3
+  namespace: terraform-system
+spec:
+  components:
+    - name: sml-aws-s3
+      type: aws-s3
+      properties:
+        bucket: "sml-aws-s3"
+        acl: "private"
+        region: "us-west-1"    //替换provider默认的region
+        providerRef:    //指定provider
+          name: aws
+          namespace: default
+        writeConnectionSecretToRef:    //存储s3 url
+          name: "sml-aws-s3-secret"
+          namespace: terraform-system
+```
+
+kubectl apply即可
+
+如果用户不知道aws-s3 component格式，可以使用命令查看：
+
+```
+#vela show aws-s3
+```
+
+详细component definition，请参考：
+
+```
+#kubectl get componentdefinition aws-s3 -n vela-system -o yaml
+```
+
+
+
+### 查看创建资源
+
+#### configuration
```

Comment:
Could you please capitalize the first name for all `configuration` when it mean Terraform Configuration?